storages/backends/gcloud.py

```
@@ -0,0 +1,226 @@
+        return file_object
+
+    def _save(self, name, content):
+        cleaned_name = self._clean_name(name)
+        name = self._normalize_name(cleaned_name)
+        size = getattr(content, 'size')
+
+        content.name = cleaned_name
+        encoded_name = self._encode_name(name)
+        file = self.file_class(encoded_name, 'rw', self)
+        file.blob.upload_from_file(content, size=size)
+        return cleaned_name
+
+    def delete(self, name):
+        name = self._normalize_name(self._clean_name(name))
+        self.bucket.delete_blob(self._encode_name(name))
+
+    def exists(self, name):
+        if not name:  # root element aka the bucket
+            try:
+                self.bucket
+                return True
+            except ImproperlyConfigured:
+                return False
+
+        name = self._normalize_name(self._clean_name(name))
+        return bool(self.bucket.get_blob(self._encode_name(name)))
+
+    def listdir(self, name):
+        name = self._normalize_name(self._clean_name(name))
+        # for the bucket.list and logic below name needs to end in /
+        # But for the root path "" we leave it as an empty string
+        if name and not name.endswith('/'):
+            name += '/'
+
+        files_list = list(self.bucket.list_blobs(prefix=self._encode_name(name)))
+        files = []
+        dirs = set()
+
+        base_parts = name.split("/")[:-1]
+        for item in files_list:
+            parts = item.name.split("/")
+            parts = parts[len(base_parts):]
+            if len(parts) == 1 and parts[0]:
+                # File
+                files.append(parts[0])
+            elif len(parts) > 1 and parts[0]:
+                # Directory
+                dirs.add(parts[0])
+        return list(dirs), files
+
+    def size(self, name):
+        name = self._normalize_name(self._clean_name(name))
+        blob = self.bucket.get_blob(self._encode_name(name))
```

Ugh really do not like that `get_blob` suppresses errors, it should not do that. I'd prefer to fail loudly when we do an action on something that doesn't exist. Size of 0 doesn't make too much sense to me. Maybe we wrap it up in our own `_get_blob` and just `raise NotFound` when it's...not found? Thoughts?